sql/wsrep_mysqld.cc

```diff
@@ -1712,34 +1712,51 @@ int wsrep_to_buf_helper(
   return ret;
 }
 
-static int
-wsrep_alter_query_string(THD *thd, String *buf)
+static int wsrep_alter_event_query(THD *thd, uchar** buf, size_t* buf_len)
 {
-  /* Append the "ALTER" part of the query */
-  if (buf->append(STRING_WITH_LEN("ALTER ")))
-    return 1;
-  /* Append definer */
-  append_definer(thd, buf, &(thd->lex->definer->user), &(thd->lex->definer->host));
-  /* Append the left part of thd->query after event name part */
-  if (buf->append(thd->lex->stmt_definition_begin,
-                  thd->lex->stmt_definition_end -
-                  thd->lex->stmt_definition_begin))
+  String log_query;
+
+  /* build the  ALTER statement, with definer */
+  if (log_query.append(STRING_WITH_LEN("ALTER "))
+      || append_definer(thd, &log_query, &(thd->lex->definer->user), &(thd->lex->definer->host))
+      || log_query.append(thd->lex->stmt_definition_begin,
+                           thd->lex->stmt_definition_end -
+                           thd->lex->stmt_definition_begin))
+  {
+    WSREP_WARN("events alter string failed: schema: %s, query: %s",
+               thd->get_db(), thd->query());
     return 1;
+  }
 
-  return 0;
+  return wsrep_to_buf_helper(thd, log_query.ptr(), log_query.length(),
+                             buf, buf_len);
 }
 
-static int wsrep_alter_event_query(THD *thd, uchar** buf, size_t* buf_len)
+static int wsrep_alter_table_query(THD *thd, uchar** buf, size_t* buf_len,
+                                   Alter_info *alter_info)
 {
   String log_query;
+  log_query.append(thd->query());
 
-  if (wsrep_alter_query_string(thd, &log_query))
+  /*
+     if user has specified the alter algorithm by session variable alter_algorithm
+     and the ALTER statement does not contain ALGORITHM= clause, then
+     build for replication new ALTER query with the ALGORITHM clause
+   */
+  if (thd->variables.alter_algorithm  != Alter_info::ALTER_TABLE_ALGORITHM_DEFAULT &&
```

Comment:
Why cannot algorithm clause be always included? This way the `Alter_info::requested_algorithm` wouldn't have to be public.

Reply:
I wanted to avoid query rewrite, unless absolutely required. If rewrite causes issues, in some scenarios, it would affect all ALTER statement processing. I would expect that alter_algorithm session variable use is somewhat minimal "trick", in practice